R/data.table.R

```diff
@@ -1942,9 +1942,42 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
       X[[j]] = xj
     }
   }
-  X = unlist(X, recursive = FALSE, use.names = FALSE)
-  dim(X) <- c(n, length(X)/n)
-  dimnames(X) <- list(rownames.value, unlist(collabs, use.names = FALSE))
+  
+  # If any columns of x are multi-column, which can only occur if x is
+  # constructed incorrectly (see test 2074.07), use the old non-C
+  # method of converting to a matrix.
+  col.classes = sapply(X, class)
+  if (any(col.classes %in% c("matrix", "data.frame", "data.table"))) { 
+    X = unlist(X, recursive = FALSE, use.names = FALSE)
+    dim(X) = c(n, length(X)/n)
+    dimnames(X) = list(rownames.value, unlist(collabs, use.names = FALSE))
+    return(X)
+  } 
+  
+  # convert columns to common class before handing over to C
+  class.order = c("logical"=1, "integer"=2, "numeric"=3, "complex"=4,
```

Comment:
`init.c` uses a slightly different heirarchy:

```
static void setSizes() {
  for (int i=0; i<100; ++i) { sizes[i]=0; typeorder[i]=0; }
  // only these types are currently allowed as column types :
  sizes[LGLSXP] =  sizeof(int);       typeorder[LGLSXP] =  0;
  sizes[RAWSXP] =  sizeof(Rbyte);     typeorder[RAWSXP] =  1;
  sizes[INTSXP] =  sizeof(int);       typeorder[INTSXP] =  2;   // integer and factor
  sizes[REALSXP] = sizeof(double);    typeorder[REALSXP] = 3;   // numeric and integer64
  sizes[CPLXSXP] = sizeof(Rcomplex);  typeorder[CPLXSXP] = 4;
  sizes[STRSXP] =  sizeof(SEXP *);    typeorder[STRSXP] =  5;
  sizes[VECSXP] =  sizeof(SEXP *);    typeorder[VECSXP] =  6;   // list column
  if (sizeof(char *)>8) error(_("Pointers are %d bytes, greater than 8. We have not tested on any architecture greater than 64bit yet."), sizeof(char *));
  // One place we need the largest sizeof is the working memory malloc in reorder.c
}
```

I see you've taken your order directly from `?typeof` (or at least it matches that). However, this doesn't matched the referenced `TypeTable` in `r-base/src/main/util.c`:

```
TypeTable[] = {
    { "NULL",		NILSXP	   },  /* real types */
    { "symbol",		SYMSXP	   },
    { "pairlist",	LISTSXP	   },
    { "closure",	CLOSXP	   },
    { "environment",	ENVSXP	   },
    { "promise",	PROMSXP	   },
    { "language",	LANGSXP	   },
    { "special",	SPECIALSXP },
    { "builtin",	BUILTINSXP },
    { "char",		CHARSXP	   },
    { "logical",	LGLSXP	   },
    { "integer",	INTSXP	   },
    { "double",		REALSXP	   }, /*-  "real", for R <= 0.61.x */
    { "complex",	CPLXSXP	   },
    { "character",	STRSXP	   },
    { "...",		DOTSXP	   },
    { "any",		ANYSXP	   },
    { "expression",	EXPRSXP	   },
    { "list",		VECSXP	   },
    { "externalptr",	EXTPTRSXP  },
    { "bytecode",	BCODESXP   },
    { "weakref",	WEAKREFSXP },
    { "raw",		RAWSXP },
    { "S4",		S4SXP },
    /* aliases : */
    { "numeric",	REALSXP	   },
    { "name",		SYMSXP	   },

    { (char *)NULL,	-1	   }
};
```

Not sure what to do here, exactly 🤔

Reply:
Anyway, I think you don't want to use `class` here -- you want to use `typeof`, since what you're after is the storage type

Reply:
Hmmm, I think the only one out of order is type "raw", but testing with unlist and as.matrix.data.frame shows that if there are any raw byte columns, the atomic type of the matrix should be character

Reply:
I'd also be tempted to revert to the unlist() approach if types not listed are detected; e.g. in case there are people doing something exotic like have columns containing environments or functions that would currently crash.

Reply:
Revisiting this - when coercing columns we do want the class not the storage type. An example why is illustrated by the bit64::integer64 class. Its type is simply a double, but if we had a data.table with a mix of logical, integer, and integer64 columns, then if we coerced based on type, the logical and integer columns would become numerics. This is what the `unlist` function does, stripping the `class` from the integer64 column leaving nonsensical values. If we instead subsequently set the class of the matrix to integer64, then the integer64 column would be correct, but the other two columns would have nonsense values. By detecting class, we can instead coerce these using as.integer64, correctly obtaining an integer64 matrix.

I'm not sure how to properly handle this more generally. As I noted in the other review thread below, the `as()` function doesn't seem to handle class coercion reliably, so it seems we have to implement this conversion manually for each new class of interest. At this time, if we detect columns that are not one of the base R atomic classes or integer64 we fallback on the `unlist()` method (unless all columns are the same type, in which case the class information is simple to preserve). However, this is imperfect, because you end up with a matrix coerced based on the storage mode which is likely to lead to nonsense values.

Reply:
I've split out the type and class coercion, with coercion based on storage type occurring as the fallback method when there are multiple columns with different classes that haven't otherwise been handled.

Reply:
Noting here now that i've added support for raw matrices: the type coercion rules I've applied to raw differ from init.c listed above. As far as I can tell, raws should be coerced to logicals, not the other way round, e.g. this is the way `c()` handles type coercion: `c(as.raw(1), TRUE) == c(TRUE, TRUE)`. I think this is because raw vectors cannot contain `NA` while logical can.

Reply:
could you try adjusting the order in init.c and see what errors this causes?

Reply:
It causes test 2006.1, 2006.2, and 2129 to fail. 

Tests 2006.1 and 2006.2 are for `rbindlist()`, testing whether `fill=TRUE` works with raw. It assumes that the corresponding `NA` in the missing column is filled with as.raw(0), whereas with the order change, the resulting data.table columns are coerced to logical.

Note: 
```
> as.raw(NA)
[1] 00
Warning message:
out-of-range values treated as 0 in coercion to raw 
```

I'm not sure why test 2129 errors, as there are no raw columns in the input data.tables.